### ISSUE TITLE
Y026 `TypeAlias` check causes many false positives

### ISSUE THREAD

**Issue by AlexWaygood** [PyCQA/flake8-pyi#205]
We're now at a point where all major type checkers have support for `typing(_extensions).TypeAlias`, so we should be looking at enabling Y026 in typeshed's `.flake8` config file. But if I enable Y026 and then run flake8 on `stdlib/random.pyi`, here are the results:

```
stdlib\random.pyi:129:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:130:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:131:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:132:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:133:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:134:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:135:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:136:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:137:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:138:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:139:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:140:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:141:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:142:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:143:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:144:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:145:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:146:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:147:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:148:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:149:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:150:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\random.pyi:152:5: Y026 Use typing_extensions.TypeAlias for type aliases
```

Typeshed source code here: https://github.com/python/typeshed/blob/7f9be7fd00ff86750a3d983681a9ef94b5511e2c/stdlib/random.pyi#L128-L152

These are all false positives. `typing_extensions.TypeAlias` cannot be used for any of these assignments, because the object on the right-hand side of the assignment isn't valid as a type (and therefore cannot be validly used in a type annotation) in any of these cases. These aren't _type_ aliases, they're _variable_ aliases.

There are similar false positives in multiple other stub files in typeshed. For example, in  <details> <summary>`re.pyi` (source code here: https://github.com/python/typeshed/blob/7f9be7fd00ff86750a3d983681a9ef94b5511e2c/stdlib/re.pyi#L110-L149) </summary>

```
stdlib\re.pyi:111:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:112:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:113:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:114:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:115:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:116:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:117:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:118:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:119:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:120:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:121:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:122:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:123:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:124:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:125:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:126:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:127:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:131:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:132:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:133:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:134:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:135:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:136:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:137:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:138:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:139:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:140:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:141:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:142:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:143:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:144:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:145:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:146:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:147:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:149:5: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\re.pyi:150:1: Y026 Use typing_extensions.TypeAlias for type aliases
```

</details>

**Comment by AlexWaygood** [comment]
Two other, slightly different, false positives in `stdlib/email/base64mime.pyi`. The typeshed stub on lines 6-9 is:

```python
def decode(string: str | bytes) -> bytes: ...

body_decode = decode
decodestring = decode
```

flake8-pyi says:

```
stdlib\email\base64mime.pyi:8:1: Y026 Use typing_extensions.TypeAlias for type aliases
stdlib\email\base64mime.pyi:9:1: Y026 Use typing_extensions.TypeAlias for type aliases
```

But again, these aren't type aliases, they're variable aliases.

**Comment by srittau** [comment]
The check looks too naive at the moment. I think it will be hard to come with a good check, but some heuristics: Does the right side use a `typing` import or and pipe character. This will still miss simple assignments like `SomeAlias = OtherAlias`, but I don't think we can do much better, without analyzing more than just the current line.

**Comment by AlexWaygood** [comment]
> The check looks too naive at the moment. I think it will be hard to come with a good check, but some heuristics: Does the right side use a `typing` import or and pipe character. This will still miss simple assignments like `SomeAlias = OtherAlias`, but I don't think we can do much better, without analyzing more than just the current line.

I have a few ideas -- I'm playing around with a patch now :)

**Comment by JelleZijlstra** [comment]
I feel like this is better handled in a type checker. Mypy and pyright must already know when they're inferring a type alias, so they should be able to give an error if they see one without `: TypeAlias`. That way, we don't have to come up with our own heuristics.

**Comment by AlexWaygood** [comment]
> I feel like this is better handled in a type checker. Mypy and pyright must already know when they're inferring a type alias, so they should be able to give an error if they see one without `: TypeAlias`. That way, we don't have to come up with our own heuristics.

Should we just get rid of Y026, then?

**Comment by srittau** [comment]
I'd suggest to leave a (fixed) check in, until a type checker has an appropriate warning. But I agree, a type checker probably has a much better idea what a valid type alias is and what not.
